oadrtest/oadrtest/tests/VENManagerTest.cpp

```diff
@@ -0,0 +1,116 @@
+	}
+}
+
+TEST(VENManagerTest, venGetsRegistrated_vtnRespondsToFirstPollWithReregister_getIntoUnregisteredState)
+{
+	using namespace testing;
+
+	StrictMock<GMockHttp> http;
+	StrictMock<MockEventService> eventService;
+	BlankReportService reportService;
+	StrictMock<MockOADRExceptionService> exceptionService;
+
+	std::string createPartyRegistrationResponse = LoadFile::loadTestInputFile("created_party_registration.xml");
+	std::string requestReregistration = LoadFile::loadTestInputFile("request_reregistration.xml");
+
+	std::unique_ptr<IVENManager> venManager(createVENManager(http, eventService, reportService, exceptionService));
+
+	const auto stopVENManager = [&]()
+	{
+		venManager->stop();
+		return true;
+	};
+
+	// If IHttp::post() is called but none of expected calls is satisfied, let the VENManager to stop.
+	ON_CALL(http, post(_, _)).WillByDefault(stopVENManager);
+
+	// TODO: consider removing IHttp::getResponseMessage since nobody is using that
+	std::string empty = "";
+	EXPECT_CALL(http, getResponseMessage()).WillRepeatedly(ReturnRef(empty));
+
+	std::string responseCode = "200";
+	EXPECT_CALL(http, getResponseCode()).WillRepeatedly(ReturnRef(responseCode));
+
+	InSequence expectCallsInSequence;
+	// The first attempt to register the VEN.
+	{
+		// Send oadrCreatePartyRegistration and receive oadrCreatedPartyRegistration
+		EXPECT_CALL(http, post(_, HasSubstr("oadrCreatePartyRegistration"))).WillOnce(Return(true));
+		EXPECT_CALL(http, getResponseBody()).Times(AtLeast(1)).WillRepeatedly(ReturnRef(createPartyRegistrationResponse));
+
+		// Send oadrRegisterReport and receive any message (it is ignored anyway)
+		EXPECT_CALL(http, post(_, HasSubstr("oadrRegisterReport"))).WillOnce(Return(true));
+		EXPECT_CALL(http, getResponseBody()).Times(AtLeast(1)).WillRepeatedly(ReturnRef(requestReregistration));
+
+		// Send oadrPoll and receive oadrRequestReregistration
+		EXPECT_CALL(http, post(_, HasSubstr("oadrPoll"))).WillOnce(Return(true));//.WillOnce(Return(true));
+		EXPECT_CALL(http, getResponseBody()).Times(AtLeast(1)).WillRepeatedly(ReturnRef(requestReregistration));
+	}
+	// The second attempt to register the VEN.
+	{
+		// Restart the process but this time command the VENManager to stop after the process finishes
+		EXPECT_CALL(http, post(_, HasSubstr("oadrCreatePartyRegistration"))).WillOnce(stopVENManager);
+		EXPECT_CALL(http, getResponseBody()).Times(AtLeast(1)).WillRepeatedly(ReturnRef(createPartyRegistrationResponse));
+
+		EXPECT_CALL(http, post(_, HasSubstr("oadrRegisterReport"))).WillOnce(Return(true));
+		EXPECT_CALL(http, getResponseBody()).Times(AtLeast(1)).WillRepeatedly(ReturnRef(requestReregistration));
+
+		EXPECT_CALL(http, post(_, HasSubstr("oadrPoll"))).WillOnce(Return(true));
+		EXPECT_CALL(http, getResponseBody()).Times(AtLeast(1)).WillRepeatedly(ReturnRef(requestReregistration));
+	}
+
+	venManager->run();
```

Comment:
It is quite unusual that the test is blocked until some condition is met, so be warned! In case of this test, we have 2 ways to get out:
1. when everything goes nice: https://github.com/AnalyticsFire/OpenADR-VEN-Library/pull/14/files#diff-dcbbb7bd023c04f368cfa425872881c72ef66573ff28a7b853c0d7feb5c2d6fbR105
2. when things go bad: https://github.com/AnalyticsFire/OpenADR-VEN-Library/pull/14/files#diff-dcbbb7bd023c04f368cfa425872881c72ef66573ff28a7b853c0d7feb5c2d6fbR78